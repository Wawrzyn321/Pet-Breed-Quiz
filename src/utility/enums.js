const AppState = Object.freeze({
  Quiz: "QUIZ",
  Menu: "MENU"
});

const QuizState = Object.freeze({
  None: "NONE",
  ShowingQuestion: "SHOWING_QUESTION",
  ShowingAnswer: "SHOWING_ANSWER",
  Finished: "FINISHED"
});

const Categories = Object.freeze({
  Dogs: "DOGS",
  Cats: "CATS",
  Both: "BOTH",
  None: "NONE"
});

const Creatures = Object.freeze({
  Dog: "DOG",
  Cat: "CAT"
});

const AnswerState = Object.freeze({
  None: "NONE",
  Bad: "BAD",
  Good: "GOOD"
});

export { AppState, QuizState, Categories, Creatures, AnswerState };
