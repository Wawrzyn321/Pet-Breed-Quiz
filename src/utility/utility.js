const CONNECTION_TIMEOUT_MESSAGE = "Connection timeout";
const NETWORK_ERROR_MESSAGE = "NetworkError when attempting to fetch resource.";

function isConnectionError(error) {
  return (
    error.message === CONNECTION_TIMEOUT_MESSAGE ||
    error.message === NETWORK_ERROR_MESSAGE
  );
}

function getShuffledArray(sourceArray) {
  const array = [...sourceArray];
  for (let i = array.length - 1; i > 0; i--) {
    const j = getRandomNumber(i + 1);
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}

function getRandomElements(array, count) {
  if (count > array.length) {
    throw new Error(
      "getRandomIndices: count must be less or equal the array length!"
    );
  }

  const shuffled = getShuffledArray(array);
  return shuffled.splice(0, count);
}

function getRandomNumber(maxExclusive) {
  return Math.floor(Math.random() * maxExclusive);
}

function fetchTimeout(url, options, timeout = 10000) {
  return Promise.race([
    fetch(url, options),
    new Promise((resolve, reject) =>
      setTimeout(() => reject(new Error(CONNECTION_TIMEOUT_MESSAGE)), timeout)
    )
  ]);
}

export {
  isConnectionError,
  getShuffledArray,
  getRandomElements,
  getRandomNumber,
  fetchTimeout
};
