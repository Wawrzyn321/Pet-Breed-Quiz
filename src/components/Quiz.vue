<template id="quiz-template">
  <div id="quiz" :class="containerClass">
    <div>
      <quiz-title-component
        :question-index="currentQuestionIndex"
        :questions-count="questionsCount"
      ></quiz-title-component>

      <quiz-question-component
        :current-question="currentQuestion"
        v-on:answer-marked="onAnswerMarked"
        ref="questionDisplay"
      >
      </quiz-question-component>

      <quiz-control-component
        :state="quizState"
        :quiz-score="score"
        :questions-count="questionsCount"
        v-on:finish-quiz-requested="finishQuiz"
        v-on:next-question-requested="showNextQuestion"
      >
      </quiz-control-component>
    </div>
  </div>
</template>

<script>
import QuizCreator from "./../QuizCreator";
import QuizTitleComponent from "./QuizTitle.vue";
import QuizQuestionComponent from "./QuizQuestion.vue";
import QuizControlComponent from "./QuizControl.vue";
import {
  Categories,
  AppState,
  QuizState,
  AnswerState
} from "./../utility/enums";
import { isConnectionError } from "./../utility/utility";

export default {
  data() {
    return {
      answersCount: 4,
      questionsCount: 10,

      choosenCategory: Categories.None,

      questions: [],
      quizState: QuizState.None,
      currentQuestionIndex: -1,
      score: 0
    };
  },
  computed: {
    containerClass() {
      return {
        "non-visible": this.appState !== AppState.Quiz
      };
    },
    currentQuestion() {
      if (this.questions[this.currentQuestionIndex] !== undefined) {
        return this.questions[this.currentQuestionIndex];
      } else {
        return null;
      }
    }
  },
  methods: {
    async initializeQuiz(choosenCategory) {
      try {
        const questions = await QuizCreator.create(
          choosenCategory,
          this.questionsCount,
          this.answersCount
        );
        this.questions = JSON.parse(JSON.stringify(questions)); // it doesn't work the simple '=' way
      } catch (err) {
        if (isConnectionError(err)) {
          alert(
            "It seems that one of our pet providers is out! :( Is your Internet connection okay by the way?"
          );
          this.finishQuiz();
        } else {
          throw err;
        }
      }

      this.score = 0;
      this.currentQuestionIndex = 0;
      this.quizState = QuizState.ShowingQuestion;
    },

    onAnswerMarked(answer) {
      // check if user has already marked the answer
      if (this.quizState !== QuizState.ShowingQuestion) {
        return;
      }

      const choosenAnswerIndex = this.currentQuestion.answers.findIndex(
        el => el.displayName === answer.displayName
      );

      if (choosenAnswerIndex === this.currentQuestion.answerIndex) {
        this.currentQuestion.answers[choosenAnswerIndex].state =
          AnswerState.Good;
        this.score++;
      } else {
        this.currentQuestion.answers[choosenAnswerIndex].state =
          AnswerState.Bad;
        this.currentQuestion.answers[this.currentQuestion.answerIndex].state =
          AnswerState.Good;
      }

      if (this.currentQuestionIndex === this.questionsCount - 1) {
        this.quizState = QuizState.Finished;
      } else {
        this.quizState = QuizState.ShowingAnswer;
      }
    },

    showNextQuestion() {
      this.currentQuestionIndex++;
      this.quizState = QuizState.ShowingQuestion;
    },

    finishQuiz() {
      this.reset();
      this.$emit("quiz-finished");
    },

    reset() {
      this.quizState = QuizState.None;
      this.choosenAnswerIndex = -1;
      this.score = 0;
      this.questions = [];
      this.$refs.questionDisplay.clearImage();
    }
  },
  components: {
    QuizTitleComponent,
    QuizQuestionComponent,
    QuizControlComponent
  },
  props: {
    appState: {
      required: true,
      type: String
    }
  }
};
</script>

<style lang="scss" scoped>
#quiz {
  display: flex;
  justify-content: center;
  position: absolute;

  top: 0;
  left: 0;
  right: 0;
  min-width: 400px;

  transition: opacity 1s;
}

#quiz.non-visible {
  visibility: hidden;
  opacity: 0;
  transition: opacity 1s;
}
</style>
