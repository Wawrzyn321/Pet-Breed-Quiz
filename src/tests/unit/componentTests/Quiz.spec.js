import Quiz from "@/components/Quiz";
import { shallowMount, mount } from "@vue/test-utils";
import { QuizState, AppState } from "@/utility/enums";

describe("Quiz", () => {
  it("currentQuestion returns current question when questions are provided", () => {
    const wrapper = shallowMount(Quiz, {
      propsData: { appState: AppState.Quiz }
    });
    wrapper.setData({
      questions: [{ a: "a" }, { b: "b" }],
      currentQuestionIndex: 1
    });

    expect(wrapper.vm.currentQuestion).toEqual({ b: "b" });
  });

  it("currentQuestion returns null when questions are not provided", () => {
    const wrapper = shallowMount(Quiz, {
      propsData: { appState: AppState.Quiz }
    });
    wrapper.setData({
      questions: [],
      currentQuestionIndex: 0
    });

    expect(wrapper.vm.currentQuestion).toBeNull();
  });

  it("changes state to finished when there are no more questions", () => {
    const mockQuestions = [
      {
        answers: [{ displayName: "fake display name" }],
        answerIndex: 0
      }
    ];
    const wrapper = shallowMount(Quiz, {
      propsData: { appState: AppState.Quiz }
    });
    wrapper.setData({
      currentQuestionIndex: 0,
      questions: mockQuestions,
      quizState: QuizState.ShowingQuestion,
      questionsCount: mockQuestions.length
    });

    wrapper.vm.onAnswerMarked(mockQuestions[0].answers[0]);

    expect(wrapper.vm.quizState).toBe(QuizState.Finished);
  });

  it("changes state to showing question when there more questions", () => {
    const mockQuestions = [
      {
        answers: [{ displayName: "fake display name 1" }],
        answerIndex: 0
      },
      {
        answers: [{ displayName: "fake display name 2" }],
        answerIndex: 0
      }
    ];
    const wrapper = shallowMount(Quiz, {
      propsData: { appState: QuizState.ShowingQuestion }
    });
    wrapper.setData({
      currentQuestionIndex: 0,
      questions: mockQuestions,
      quizState: QuizState.ShowingQuestion,
      questionsCount: mockQuestions.length
    });

    wrapper.vm.onAnswerMarked(mockQuestions[0].answers[0]);

    expect(wrapper.vm.quizState).toBe(QuizState.ShowingAnswer);
  });

  it("ignores marking answer when other answer has already been marked", () => {
    const wrapper = shallowMount(Quiz, {
      propsData: { appState: AppState.Quiz }
    });
    wrapper.setData({
      quizState: QuizState.ShowingAnswer
    });

    wrapper.vm.onAnswerMarked();

    expect(wrapper.vm.quizState).toBe(QuizState.ShowingAnswer);
  });

  it("emits quiz-finished event when quiz is finished", () => {
    const wrapper = mount(Quiz, {
      propsData: { appState: AppState.Quiz }
    });

    wrapper.vm.finishQuiz();

    expect(wrapper.emitted()).toEqual({ "quiz-finished": [[]] });
  });
});
