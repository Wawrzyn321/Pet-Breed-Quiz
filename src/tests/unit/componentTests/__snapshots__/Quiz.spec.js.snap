// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz renders with minimal props 1`] = `
<div
  class="non-visible"
  id="quiz"
>
  <div>
    <quiz-title-component-stub
      questionindex="-1"
      questionscount="10"
    />
     
    <quiz-question-component-stub />
     
    <quiz-control-component-stub
      questionscount="10"
      quizscore="0"
      state="NONE"
    />
  </div>
</div>
`;
